Update the checkpointing guide with Orbax migration code examples

IvyZX opened the pull request.

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -1,67 +1,85 @@
 ---
 
-# Checkpointing with `flax.training.checkpoints`
+# Save / Load Checkpoints
```
Nit:

```suggestion
# Save / load checkpoints
```

or "Save and load..."

(with "sentence case" https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings)

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -1,67 +1,85 @@
 
 In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, autom ...
 
+> **_Note:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to [Orbax](https://github.com/google/orbax), which provides a more customizable and flexible API for various array types and storage formats. All existed features of Flax will continue to be supported, though the API may be slightly changed. This page will contain the most up-to-date code exampl ...
```
Suggestions:

- "All existing features in the Flax API will continue to be supported, though the API may change."
- "Once the migration is complete, this guide will contain the most up-to-date code examples with Orbax-Flax checkpointing."

```suggestion
> **_Note:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to [Orbax](https://github.com/google/orbax), which provides a more customizable and flexible API for various array types and storage formats. All existing features in the Flax API will continue to be supported, though the API may change. This page will contain the most up-to-date code examples to use Orbax-Flax checkpointing.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -1,67 +1,85 @@
 Create a pytree with many data structures and containers, and play with it:
 
-```{code-cell} ipython3
+```python
````
This is cool. If the `jupytext` pre-commit test passes, we can manually change/improve all our other Markdown files.

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -70,11 +88,13 @@ variables = model.init(key2, x1)
     apply_fn=model.apply,
     params=variables['params'],
     tx=tx)
+# Make some simple fake grad update like the ones done in normal training.
```
Suggestion:

Since it's only a forward pass and the example in the guide is just a toy example, then maybe:

```suggestion
# Perform a simple gradient update similar to the one during a normal training workflow.
```

(Internally, `.apply_gradients` Flax and Optax (`tx`) code, isn't it?)

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -1,67 +1,85 @@
 ```
 
+```python
+# Orbax needs to enable asyncio in colab environment.
````
Nit:

```suggestion
# Orbax needs to enable asyncio in a Colab environment.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -1,67 +1,85 @@
-# Checkpointing with `flax.training.checkpoints`
+# Save / Load Checkpoints
 
 In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, autom ...
```
Suggestion:

- Maybe it's worth adding a sentence about Orbax - `orbax.Checkpointer` - which the user can use with `flax.checkpoints.save_checkpoint`, as demonstrated later the "Save checkpoints" section.
- Later, we can rephrase the entire introduction after the Orbax migration is complete.

For example:

```suggestion
In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.

In addition, you will be using the [Orbax](https://github.com/google/orbax) Checkpointer. Refer to the note about Orbax below.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -1,67 +1,85 @@
+```python
+!pip install -U -q flax jax jaxlib optax tensorstore orbax
+
+# Installing Flax and Orbax from head to use latest checkpoint features.
+# This part will be removed after both make a new release.
````
Suggestion:

```suggestion
# To use the latest checkpoint features, install Flax and Orbax from head.
# Note: This section will be removed/replaced after the new release for both libraries.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +221,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
+You can do this by creating an `orbax.AsyncCheckpointer` (as demonstrated in the code cell below) and let it track your save thread.
 
-`async_manager` is a parameter in [`flax.training.checkpoints.save_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint) with the default setting at `None`.
+Note that you should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This allows bookkeeping like `keep` and `overwrite` to be consistent across steps.
```
Suggestion:

```suggestion
Note: You should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This enables bookkeeping, such as `keep` and `overwrite` to be consistent across steps.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +221,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
+In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data arr ...
 
-Flax provides an easy interface for users to pass in a `GlobalAsyncCheckpointManager` and store pytrees with multi-process arrays in the same fashion as single-process pytrees. Just use [`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess) with the same arguments.
+Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, we recommend using the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can do computation along side the saves. 
```
Nit:

```suggestion
Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, it's recommended using the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can perform computation alongside the saves. 
```

(First person like "we" and "our" -> second person https://developers.google.com/style/person though "we" sounds OK, IMHO)

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +221,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
+Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, we recommend using the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can do computation along side the saves. 
 
-Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. As a workaround, use the following code as a sample to run your multi-host checkpointing.
+Code wise, just use [`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess) with the same arguments.
```
Nit: What does "code-wise" mean here? Maybe we can say:

```
In your code, use `flax.training.checkpoints.save_checkpoint_multiprocess`...
...
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -193,46 +221,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
-# Refer to https://jax.readthedocs.io/en/latest/multi_process.html#initializing-the-cluster
-jax.distributed.initialize("localhost:8889", num_processes=1, process_id=0)
+```python
+# Multihost related imports.
````
Nit: "Multi-host" ?

8bitmp3 reviewed (commented):
An awesome update to the guide @IvyZX 👍 

A few suggestions/recommendations:

- In the intro, maybe it's worth adding a sentence about Orbax - `orbax.Checkpointer` - before your note about the migration. The Orbax API is demonstrated with the Flax API in the "Save checkpoints" section (new content). In future, we can rephrase the entire introduction after the Orbax migration is 100% complete.
- In the note: "All existing features in the Flax API will continue to be supported, though the API may change... Once the migration is complete, this guide will contain the most up-to-date code examples with Orbax-Flax checkpointing."
- "# To use the latest checkpoint features, install Flax and Orbax from head."
  "# Note: This section will be removed/replaced after the new release for both libraries."
- In the `.apply_gradients` step: "Since it's only a forward pass and the example in the guide is just a toy example, then maybe: # Perform a simple gradient update similar to the one during a normal training workflow."
- And a few other minor suggestions.

LMKWYT!

andsteing commented:
In the `Setup` part, could we get rid of all this noisy `pip install` output?
(Consider using `-qq`, and/or clearing the output cell)

IvyZX commented:
> Did you try using `chex.set_n_cpu_devices()` instead of fiddling with `--xla_force_host_platform_device_count` directly?

`xla_force_host_platform_device_count`  is the recommended way to emulate multiple devices on CPU in [JAX doc](https://jax.readthedocs.io/en/latest/jax-101/06-parallelism.html#aside-hosts-and-devices-in-jax). 

On the other hand, chex is not an official dependency of Flax and was not used in the codebase other than in `jax_utils_test.py`, so I think it's better to not introduce extra dependency in this doc.

IvyZX commented:
Thank you Olzhas and Andreas for all the reviews! I have address them all in the new merged code. PTAL.

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -1,20 +1,23 @@
+# Save and load checkpoints
 
-In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, autom ...
+In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints) — a simplistic and generic checkpointing library built into Flax. Flax allows you to save and load model parameters, metadata, and a variety of Python data using the library [Orbax](https://github.com/google/orbax), which provid ...
```
Since checkpointing will be taken care of by Orbax, maybe we can omit `flax.training.checkpoints` now.

For example:

```suggestion
In this guide, you will learn about saving and loading checkpoints with Flax and Orbax](https://github.com/google/orbax). Flax allows you to save and load model parameters, metadata, and a variety of Python data using Orbax, which provides a more customizable and flexible API for various array types and storage formats. In addition, Flax provides basic features for versioning, automatic bookkeeping of past checkpoints, and asynchronous saving to reduce training wait time.
```

Then, in the **Ongoing migration** note, we can add that:

```diff
**_Ongoing migration:_** In the foreseeable future, Flax's checkpointing
functionality will gradually be migrated to Orbax
+ from `flax.training,checkpoints`.

```

Does this make sense? LMKWYT :+1:

IvyZX replied:
Sounds good! Done.

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -23,45 +26,60 @@ This guide covers the following:
-```{code-cell} ipython3
-!pip install -U -q flax jax jaxlib optax tensorstore
+```python
+!pip install -U -qq flax jax jaxlib optax tensorstore orbax
````
If the `flax` package contains JAX, Optax, etc, are we instructing users to _then upgrade_ `jax`/`jaxlib` and `optax`? Otherwise, maybe `!pip install -U -q flax tensorstore orbax` might be enough?

IvyZX replied:
Done.

levskaya commented on docs/guides/use_checkpointing.ipynb:
```diff
@@ -5,17 +5,25 @@
    "id": "6e9134fa",
```
<p><code>flax.training,checkpoints</code> -&gt; <code>flax.training.checkpoints</code> (comma to period)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/2688/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

levskaya replied:
<p>add pip quiet option to silent install warnings..... why not add this install at the very top with the rest??</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/flax/pull/2688/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -23,45 +26,58 @@ This guide covers the following:
 import optax
 ```
 
+```python
````
If possible, maybe we can suggest this step after `pip install flax` since this is still in the "Setup" stage of the guide.

For example:

```
!pip install -U -qq flax orbax
...

# Orbax requires asyncio to be insabled in a Google Colab environment
!pip install nest_asyncio
```

Then in the next code cell:

```
import jax
...
import optax
...
import nest_asyncio
nest_asyncio.apply()
```

LMKWYT

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -1,20 +1,23 @@
-In this guide, you will learn about [`flax.training.checkpoints`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#module-flax.training.checkpoints)—a simplistic and generic checkpointing library built into Flax. With Flax Checkpoints, you can save and load model parameters, metadata, and a variety of Python data. In addition, it provides basic features for versioning, autom ...
+In this guide, you will learn about saving and loading checkpoints with Flax and [Orbax](https://github.com/google/orbax). Flax allows you to save and load model parameters, metadata, and a variety of Python data using the Orbax, which provides a more customizable and flexible API for various array types and storage formats. In addition, Flax provides basic features for versioning, automatic book ...
+
+> **_Ongoing migration:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to Orbax from `flax.training,checkpoints`. All existed features in the Flax API will continue to be supported, though the API will change. You are encouraged to try out the new API by creating an `orbax.Checkpointer` and pass it in your Flax API calls as arg `orbax_checkpointer`. Yo ...
```
Nit:

How about:

```suggestion
> **_Ongoing migration:_** In the foreseeable future, Flax's checkpointing functionality will gradually be migrated to Orbax from `flax.training.checkpoints`. All existing features in the Flax API will continue to be supported, but the API will change. You are encouraged to try out the new Orbax API by creating an `orbax.Checkpointer` and pass it in your Flax API calls as an argument `orbax_checkpointer`, as demonstrated later in this guide. This guide provides the most up-to-date code examples for using Orbax and Flax for checkpointing.
```

WDYT?

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -88,29 +106,35 @@ Now save the checkpoint. You can add annotations like step number, prefix, and s
-To restore a checkpoint, use [`flax.training.checkpoints.restore_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.restore_checkpoint) and pass in the checkpoint directory. Flax will automatically select the latest checkpoint in the directory. You can also choose to specify a step number or the path of the checkpoint file. You can always ...
+To restore a checkpoint, use [`flax.training.checkpoints.restore_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.restore_checkpoint) and pass in the checkpoint directory. Flax will automatically select the latest checkpoint in the directory. You can also choose to specify a step number or the path of the checkpoint file.
+
+With the Orbax migration on progress, `restore_checkpoint` can automatically identify whether a checkpoint is saved in legacy or orbax version, and restore the pytree correctly.
```
Suggestion:

```suggestion
With the migration to Orbax in progress, `restore_checkpoint` can automatically identify whether a checkpoint is saved in the legacy (Flax) or Orbax version, and restore the Pytree correctly.
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -88,29 +106,35 @@ Now save the checkpoint. You can add annotations like step number, prefix, and s
+
+With the Orbax migration on progress, `restore_checkpoint` can automatically identify whether a checkpoint is saved in legacy or orbax version, and restore the pytree correctly.
+
+You can always restore a pytree out of your checkpoints by setting `target=None`.
```
```suggestion
You can always restore a Pytree out of your checkpoints by setting `target=None`.
```

(Maybe it's best to double check the spelling on JAX's site. Sometimes it's "PyTree" :thinking: )

IvyZX replied:
JAX site mostly uses "pytree": https://jax.readthedocs.io/en/latest/search.html?q=PyTree# 
Making it consistent here.

8bitmp3 replied:
Excellent, thanks!

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +219,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
+Note: You should use the same `async_checkpointer` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This enables bookkeeping, such as `keep` and `overwrite` to be consistent across steps.
 
-Note that you should use the same `async_manager` to handle all your async saves across your training steps, so that it can make sure that a previous async save is done before the next one begins. This allows bookkeeping like `keep` and `overwrite` to be consistent across steps.
+Whenever you want to explicitly wait until an async save is done, you can call `async_checkpointer.wait_until_finished()`. Alternatively, you can pass in `orbax_checkpointer=async_checkpointer` when running `restore_checkpoint` and Flax will automatically wait and restore safely.
```
Suggestion: Add a link to the API docs for `async_checkpointer.wait_until_finished()`.

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +219,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
 JAX provides a few ways to scale up your code on multiple hosts at the same time. This usually happens when the number of devices (CPU/GPU/TPU) is so large that different devices are managed by different hosts (CPU). To get started on JAX in multi-process settings, check out [Using JAX in multi-host and multi-process environments](https://jax.readthedocs.io/en/latest/multi_process.html).
 
-In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data arr ...
+In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data arr ...
```
Maybe we should rename `gda_serialization` -> `array_serialization` (Reference: @yashk2810 's PR `jax.Array` github.com/google/flax/pull/2703)

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -193,46 +219,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
+In the [Single Program Multi Data (SPMD)](https://jax.readthedocs.io/en/latest/glossary.html#term-SPMD) paradigm with JAX [`pjit`](https://jax.readthedocs.io/en/latest/jax.experimental.pjit.html), a large multi-process array can have its data sharded across different devices (check out the `pjit` [JAX-101 tutorial](https://jax.readthedocs.io/en/latest/jax-101/08-pjit.html)). This kind of data arr ...
 
-Flax provides an easy interface for users to pass in a `GlobalAsyncCheckpointManager` and store pytrees with multi-process arrays in the same fashion as single-process pytrees. Just use [`flax.training.checkpoints.save_checkpoint_multiprocess`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint_multiprocess) with the same arguments.
+Orbax uses this API to support storing pytrees with multi-process arrays in the same fashion as single-process pytrees. However, it's recommended to use the asynchronized `orbax.AsyncCheckpointer` to save large multi-process arrays on another thread, so that you can perform computation alongside the saves. 
```
Double check if it's "Pytree" or "PyTree" (I think it's the former).

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -193,46 +219,52 @@ serialization.from_state_dict(custom_target, raw_state_dict)
 
-```{code-cell} ipython3
-# Set up a checkpoint with a multi-process array.
+Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. Please simply use the following code as a sample to run your multi-host checkpointing.
````
Suggestion:

```suggestion
Unfortunately, Python Jupyter notebooks are single-host only and cannot activate the multi-host mode. You can treat the following code as an example for running your multi-host checkpointing:
```

8bitmp3 commented on docs/guides/use_checkpointing.md:
````diff
@@ -88,29 +106,35 @@ Now save the checkpoint. You can add annotations like step number, prefix, and s
 When saving a checkpoint, Flax will bookkeep the existing checkpoints based on your arguments. For example, by setting `overwrite=False` in [`flax.checkpoints.save_checkpoint`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.checkpoints.save_checkpoint), Flax will not automatically save your checkpoint if there is already a step that is equal to or newer than  ...
 
-```{code-cell} ipython3
+You can start to use Orbax to handle the underlying save by creating an `orbax.Checkpointer`, and pass it into the `flax.checkpoints.save_checkpoint` call.
````
If we haven't done this already, maybe we can include a link to `orbax.Checkpointer` API docs if they exist.

8bitmp3 reviewed (commented):
Thanks @IvyZX ! 

A few more minor suggestions:

- Maybe we should rename `gda_serialization` -> `array_serialization` (Reference: @yashk2810 's PR `jax.Array` github.com/google/flax/pull/2703).
- Add a link to the API docs for `async_checkpointer.wait_until_finished()`.
- Minor refactoring:

```
With the migration to Orbax in progress, `restore_checkpoint` can
automatically identify whether a checkpoint is saved in the legacy
(Flax) or Orbax version, and restore the Pytree correctly.
```

- If we haven't done this already, maybe we can include a link to `orbax.Checkpointer` API docs if they exist.
- Minor refactoring:

```
> **_Ongoing migration:_** In the foreseeable future, Flax's 
> checkpointing functionality will gradually be migrated to Orbax from 
> `flax.training.checkpoints`. All existing features in the Flax API 
> will continue to be supported, but the API will change. You are 
> encouraged to try out the new Orbax API by creating an
> `orbax.Checkpointer` and pass it in your Flax API calls as an
> argument `orbax_checkpointer`, as demonstrated later in this guide. 
> This guide provides the most up-to-date code examples for using 
> Orbax and Flax for checkpointing.
```

LMKWYT!

IvyZX commented:
Thanks for the review! Most requests are addressed.

As for `async_checkpointer.wait_until_finished`, I added the link to `AsyncCheckpointer`'s definition, which has a docstring that explains why and when user should call `wait_until_finished`. This should be a good entry point for any user that want to dig into Orbax code.

About gda_serialization -> array_serialization, I noticed we don't need to import it any more because Orbax will handle the serialization. I modified the text and imports accordingly.

8bitmp3 commented on docs/guides/use_checkpointing.md:
```diff
@@ -121,30 +140,31 @@ However, when using `target=None`, the restored `raw_restored` will be different
 ### Backward/forward dataclass compatibility
 
-The flexibility of using *Flax dataclasses*—[`flax.struct.dataclass`](https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html#flax.struct.dataclass)—means that changes in Flax dataclass fields may break your existing checkpoints. For example, if you decide to add a field `batch_stats` to your `TrainState`, old checkpoints without this field may not be successfully restored. Same goes ...
+The flexibility of using *Flax dataclasses*—[`flax.struct.dataclass`](https://flax.readthedocs.io/en/latest/api_reference/flax.struct.html#flax.struct.dataclass)—means that changes in Flax dataclass fields may break your existing checkpoints. For example, if you decide to add a field `batch_stats` to your `TrainState` (like when using [batch normalization](https://flax.readthedocs.io/en/latest/gu ...
```
fyi @cgarciae @marcvanzee we (@IvyZX and I) decided to reference the BatchNorm guide here :+1:)
